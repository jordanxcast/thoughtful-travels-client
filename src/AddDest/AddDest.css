.AddDest {
  position: relative;
  top: 100px;
}

.form-group{
  width: 60%;
  margin: auto;
}

.AddDest-title,
.AddDest-date,
.AddDest-budget {
  display: flex;
  flex-direction: row;
  margin-bottom: 10px;
  justify-content: center;
}

.add-dest-button-group > button{
  margin: 30px 20px 0 20px;
  padding: 5px 10px;
  border: 1px solid #131B23;
  border-radius: 20px;
}

.cancel-add-dest-button:hover {
  color: red;
  cursor: pointer;
}

.add-dest-button:hover {
  cursor: pointer;
  color: #A8BCA1;
}