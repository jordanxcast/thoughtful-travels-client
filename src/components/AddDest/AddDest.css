.AddDest-container {
  background-color: #4c5c68;
  height: fit-content;
  min-height: 100vh;
  padding: 0px 0px 100px;
}

.AddDest {
  position: relative;
  top: 100px;
  box-shadow: 0px 0px 8px 2px rgba(0, 0, 0, 0.7);
  justify-self: center;
  padding: 40px 30px;
  margin: 0px 20px;
  border-radius: 0px;
  border: none;
  background-color: #eafffd;
}

.AddDest-form > h2 {
  margin: 0px 0px 30px;
}

.AddDest-form-group {
  text-align: center;
  margin: auto;
  align-items: center;
}

.AddDest-title,
.AddDest-date,
.AddDest-budget {
  display: flex;
  flex-direction: column;
  margin: 0 auto 20px auto;
  align-items: center;
  width: 100%;
}
.AddDest-title > input,
.AddDest-date > input,
.AddDest-budget > input {
  height: 40px;
  background-color: #4c5c68;
  color: white;
  font-size: 16px;
  width: 100%;
  margin: auto;
}

.AddDest-title > label,
.AddDest-date > label,
.AddDest-budget > label {
  height: 40px;
  vertical-align: middle;
  line-height: 40px;
  width: auto;
  margin: 0px auto 0px;
  text-align: center;
}

.add-dest-button-group {
  display: flex;
  justify-content: space-between;
}

.add-dest-button-group > button {
  margin: 30px 20px 0 20px;
  padding: 5px 10px;
  border: 1px solid #131b23;
  border-radius: 20px;
  background: transparent;
}

.cancel-add-dest-button:hover {
  color: red;
  border-color: red;
  cursor: pointer;
}

.add-dest-button:hover {
  cursor: pointer;
  color: #a8bca1;
  border-color: #a8bca1;
}

@media (min-width: 750px) {
  .AddDest-container {
    background-color: #4c5c68;
    height: fit-content;
    min-height: 100vh;
  }
  .AddDest-title,
  .AddDest-date,
  .AddDest-budget {
    display: flex;
    flex-direction: row;
    margin: 0 auto 20px auto;
    justify-content: center;
    width: 100%;
  }
  .AddDest-title > input,
  .AddDest-date > input,
  .AddDest-budget > input {
    height: 40px;
    background-color: #4c5c68;
    color: white;
    width: 50%;
  }

  .AddDest-title > label,
  .AddDest-date > label,
  .AddDest-budget > label {
    height: 40px;
    vertical-align: middle;
    line-height: 40px;
    width: fit-content;
    text-align: right;
  }
}

@media (min-width: 500px) {
  .AddDest-title,
  .AddDest-date,
  .AddDest-budget {
    display: flex;
    flex-direction: column;
    margin: 0 auto 20px auto;
    justify-content: center;
    width: 100%;
  }

  .AddDest-title > input,
  .AddDest-date > input,
  .AddDest-budget > input {
    height: 40px;
    background-color: #4c5c68;
    color: white;
  }

  .AddDest-title > label,
  .AddDest-date > label,
  .AddDest-budget > label {
    height: 40px;
    vertical-align: middle;
    line-height: 40px;
    width: fit-content;
    text-align: right;
  }
}
