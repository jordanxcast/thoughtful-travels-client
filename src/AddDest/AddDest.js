import React, { Component } from 'react'
import ApiContext from '../ApiContext'
import config from '../config'
import TokenService from '../services/token-service'

import './AddDest.css'

class AddDest extends Component {
  static contextType = ApiContext
  constructor(props){
    super(props)
    this.state = {
      error: null
    }
  }

  validateDestTitle(value) {
    const title = value.trim()
    if(title.length === 0) {
      return 'A Destination Title is required'
    } else {
      return false
    }
  }

  validateDestDate(value) {
    if(value === '') {
      return null
    } return value
  }

  validateDestBudget(value) {
    if(value === '') {
      return 0
    } return value
  }

  clearError = () => {
    this.setState({
      error: null
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()

    const destTitle = e.target.destName.value
    const destDate = e.target.destDate.value
    const destBudget = e.target.destBudget.value

    const error = this.validateDestTitle(destTitle)
    const budget = this.validateDestBudget(destBudget)
    const goal_date =this.validateDestDate(destDate)
    
    if(error) {
      this.setState({
        error
      })
    } else {
      //dest_title, goal_date, budget
      const addDestEndpoint = `${config.API_ENDPOINT}/destinations`
      const options = {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          'authorization': `bearer ${TokenService.getAuthToken()}`
        },
        body: JSON.stringify({
          //destination inputs
          dest_title: destTitle,
          goal_date: goal_date,
          budget: budget,
        }),
      }
      fetch(addDestEndpoint, options)
      .then(res => {
        if(!res.ok) {
          throw new Error('Something went wrong, please try again later')
        } 
        return res.json()
      })
      .then(dest => {
        const destId = dest.dest_id;
        console.log('dest ', dest)
        this.context.addDest(dest)
        this.props.history.push(`/destinations/${destId}`)
      })
      .catch(err => console.error(err.message))
    }
  }

  render() {
    return (
      <div className='AddDest'>
        <form
          className='AddDest-form'
          onSubmit={(e) => this.handleSubmit(e)}
        >
          <h2>Add Your Next Destination</h2>
          <div className='form-group'>
            <label htmlFor='dest-name'>Destination Title:</label>
            <input className='dest-name' id='dest-name' name='destName' type='text' required/>
            {/* {this.state.error && (<ValidationError message={this.state.error} clearError={this.clearError}/>)} */}
            <label htmlFor='dest-date'>Goal Date</label>
            <input className='dest-date' id='dest-date' name='destDate' type='date'/>
            <label htmlFor='dest-budget'>Budget</label>
            <input className='dest-budget' id='dest-budget' name='destBudget' type='number'/>
          </div>

          <div className='add-dest-button-group'>
            <button
              type='reset'
              className='add-dest-button'
              onClick={() => this.props.history.push('/destinations')}
            >
              Cancel
            </button>
            <button
              type='submit'
              className='add-dest-button'
              disabled={this.state.error}
            >
              Save Destination
            </button>
          </div>
        </form>
      </div>
    )
  }
}

export default AddDest