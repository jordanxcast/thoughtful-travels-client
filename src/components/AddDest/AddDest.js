import React, { Component } from 'react'
import ApiContext from '../../ApiContext'
import config from '../../config'
import TokenService from '../../services/token-service'
import './AddDest.css'

class AddDest extends Component {
  static contextType = ApiContext
  constructor(props){
    super(props)
    this.state = {
      error: null
    }
  }

  validateDestTitle(value) {
    const title = value.trim()
    if(title.length === 0) {
      return 'A Destination Title is required'
    } else {
      return false
    }
  }

  validateDestDate(value) {
    if(value === '') {
      return null
    } return value
  }

  validateDestBudget(value) {
    if(value === '') {
      return 0
    } return value
  }

  clearError = () => {
    this.setState({
      error: null
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()

    const destTitle = e.target.destName.value
    const destDate = e.target.destDate.value
    const destBudget = e.target.destBudget.value

    const error = this.validateDestTitle(destTitle)
    const budget = this.validateDestBudget(destBudget)
    const goal_date =this.validateDestDate(destDate)
    
    if(error) {
      this.setState({
        error
      })
    } else {
      //dest_title, goal_date, budget
      const addDestEndpoint = `${config.API_ENDPOINT}/destinations`
      const options = {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          'authorization': `bearer ${TokenService.getAuthToken()}`
        },
        body: JSON.stringify({
          //destination inputs
          dest_title: destTitle,
          goal_date: goal_date,
          budget: budget,
        }),
      }
      fetch(addDestEndpoint, options)
      .then(res => {
        if(!res.ok) {
          throw new Error('Something went wrong, please try again later')
        } 
        return res.json()
      })
      .then(dest => {
        const destId = dest.dest_id;
        this.context.addDest(dest)
        this.props.history.push(`/destinations/${destId}`)
      })
      .catch(err => console.error(err.message))
    }
  }

  render() {
    return (
      <div className='AddDest-container'> 
        <div className='AddDest'>
          <form
            className='AddDest-form'
            onSubmit={(e) => this.handleSubmit(e)}
          >
            <h2>Add Your Next Destination</h2>
            <div className='AddDest-form-group'>
              <div className='AddDest-title'>
                <label htmlFor='dest-name'>* Destination Title:</label>
                <input className='dest-name' id='dest-name' name='destName' type='text' required/>
              </div>
            
              {/* {this.state.error && (<ValidationError message={this.state.error} clearError={this.clearError}/>)} */}
              <div className='AddDest-date'>
                <label htmlFor='dest-date'>Goal Date:</label>
                <input className='dest-date' id='dest-date' name='destDate' type='date'/>
              </div>

              <div className='AddDest-budget'>
                <label htmlFor='dest-budget'>Budget ($):</label>
                <input className='dest-budget' id='dest-budget' name='destBudget' type='number'/>
              </div>
            </div>

            <div className='add-dest-button-group'>
              <button
                type='reset'
                className='cancel-add-dest-button'
                onClick={() => this.props.history.push('/destinations')}
              >
                Cancel
              </button>
              <button
                type='submit'
                className='add-dest-button'
                disabled={this.state.error}
              >
                Save Destination
              </button>
            </div>
          </form>
        </div>
      </div>
    )
  }
}

export default AddDest