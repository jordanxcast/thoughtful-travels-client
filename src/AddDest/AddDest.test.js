import React from 'react'
import ReactDOM from 'react-dom'
import AddDest from './AddDest'
import renderer from 'react-test-renderer';


describe(`<AddDest />`, () => {
  //Smoke Testing
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <AddDest />,
      div
    );
    ReactDOM.unmountComponentAtNode(div)
  })
  //Snapshot Testing
  it('renders UI as expected', () => {
    const tree = renderer
      .create(<AddDest />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})