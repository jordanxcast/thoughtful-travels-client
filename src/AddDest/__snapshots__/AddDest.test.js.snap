// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddDest /> renders UI as expected 1`] = `
<div
  className="AddDest"
>
  <form
    className="AddDest-form"
    onSubmit={[Function]}
  >
    <h2>
      Add Your Next Destination
    </h2>
    <div
      className="form-group"
    >
      <div
        className="AddDest-title"
      >
        <label
          htmlFor="dest-name"
        >
          * Destination Title:
        </label>
        <input
          className="dest-name"
          id="dest-name"
          name="destName"
          required={true}
          type="text"
        />
      </div>
      <div
        className="AddDest-date"
      >
        <label
          htmlFor="dest-date"
        >
          Goal Date:
        </label>
        <input
          className="dest-date"
          id="dest-date"
          name="destDate"
          type="date"
        />
      </div>
      <div
        className="AddDest-budget"
      >
        <label
          htmlFor="dest-budget"
        >
          Budget:
        </label>
        <input
          className="dest-budget"
          id="dest-budget"
          name="destBudget"
          type="number"
        />
      </div>
    </div>
    <div
      className="add-dest-button-group"
    >
      <button
        className="cancel-add-dest-button"
        onClick={[Function]}
        type="reset"
      >
        Cancel
      </button>
      <button
        className="add-dest-button"
        disabled={null}
        type="submit"
      >
        Save Destination
      </button>
    </div>
  </form>
</div>
`;
